### PR TITLE
Prevent override of altitude in flightplan page

### LINKED ISSUE #928: Abnormal ALT path during approach[BUG]
**Mod Version**

**Describe the bug**

A320 FMGS generates abnormal ALT after entering APPR/STAR/TRANS on the ARRIVAL page

**To Reproduce**

1. Open Arrival page (in the example LFPG)
2. Enter APPR/STAR/TRANS (in the example=DJL8H/ILS 26L/OKIPA)
3. Open F-PLN page: an abnormally high altitude (288170 ???) was created between waypoints URELO (FL140=correct) and PG522 (9000= correct)
4. Vertical revision on OKIPA is impossible
5. Delete OKIPA waypoint is impossible

**Expected behavior**

correct path for DJL8H STAR should be TRO FL190 URELO FL140 OKIPA FL140

**Actual behavior**

Entering a STAR generates abnormal altitude at the TRANS waypoint whatever the STAR used for a given airport

**References**

FMGC2
FMGC

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### LINKED ISSUE #2261: [BUG] MCDU increasing figures replacing altitude
**Mod Version**
 "built": "2020-12-06T08:38:03+00:00",
  "ref": "",
  "sha": "",
  "actor": "lukecologne <39596827+lukecologne@users.noreply.github.com>",
  "event_name": "manual"

**Describe the bug**
enroute on the following FPL:
LLBG/12 DAFN2C DAFNA W13 VELOX M42 GENOS DCT PEDER DCT EVENO UM601
EXELA UN127 RIKSO UN128 ERESO UN127 LMO DCT EVIVI L863 OKANA DCT
SOMUN DCT PODET DCT ERKIR L608 KPT L856 RAVED DCT HOC UL856 GIVRI
UM982 TINIL TINIL9E LFPG/09R
my CRZ altitude is being replaced by strange numbers that are increasing with time (refer to picture below). These figures are spreading to the next waypoint with time too. 

image

image
Figures spreading over time ABARI still OK
image
ABARI waypoint "infected" and PENDU still OK

Note: FPL was entered manually (source Simbrief)

**To Reproduce**

1. Start C& D at LLBG (or any other airport)
2. Start-up procedure
3. Take-off and observe Altitude on F-PLN page on MCDU
4. Increasing numbers replaces step by strep over time the CRZ ALT.

**Expected behavior**
the CRZ ALT should be displayed iso increasing numbers

**Actual behavior**
Strange numbers are replacing the altitude on the F-PLN page of the MCDU

**References**
See Above 

**Additional context**
no additional context

Was this working before/when did the issue start occurring?
Couldn't say with with accuracy

Is this a problem in the vanilla unmodded game?
Unknown

Discord username (if different from GitHub):
Matt Paxton #5258

### PR DESCRIPTION
**Pull request by Saschl**
Fixes #2261 and fixes #928

## Summary of Changes

There was an issue in determining the first waypoint of the actual route (=waypoints not belonging to SID/STAR) as the waypoints are being removed step by step from the MCDU after passing it. That resulted in constantly rising numbers as the base of the prediction is the accumulated waypoint distance from the start. Therefore I added an additional check to make sure the waypoint is actually part of a STAR.

## Screenshots (if necessary)
Before:

Desktop Screenshot 2020 12 07 - 22 24 53 69 (2)

After:

Desktop Screenshot 2020 12 07 - 22 27 39 83 (2)

Discord username (if different from GitHub): saschl#9432

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Saschl** [comment]
I'll also try some cleanup if possible. The else/if blocks look somewhat redundant

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js [hunk 1]
```diff
@@ -236,6 +236,7 @@ class CDUFlightPlanPage {
                         }
                         let altitudeConstraint = "---";
                         let altPrefix = "";
+                        const isDepartureWayPoint = routeFirstWaypointIndex > 1 && mcdu.flightPlanManager.getDepartureWaypoints().indexOf(waypointsWithDiscontinuities[index]) != -1;
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:239]:
```suggestion
                        const isDepartureWayPoint = routeFirstWaypointIndex > 1 && mcdu.flightPlanManager.getDepartureWaypoints().indexOf(waypointsWithDiscontinuities[index]) !== -1;
```

**Review by MisterChocker — CHANGES_REQUESTED** [review]

**Comment by MisterChocker** [comment]
I hope these pr are not overlapping/conflicting #2245

**Comment by Saschl** [comment]
> I hope these pr are not overlapping/conflicting #2245

Dang I didn't look at this before. But at the first glance, it doesn't seem to change the logic regarding altitude constraints. As it's already Ready to test I'll wait and test after the merge again.

**Comment by beheh** [comment]
Does this also fix #928? If so, let's get it added to the PR description (with another "fixes #928").

**Comment by Saschl** [comment]
> Does this also fix #928? If so, let's get it added to the PR description (with another "fixes #928").

Looks like the same isse. I'll updated the description

**Comment by Noxleee** [comment]
Category: QA Tester
Name: Noxlee #7727
Date of testing: Current date 09/12/2020
Version of the sim: 1.11.7.0
PR Tested: #2267
Object of test: Flight plan

Testing technique: Test Flight -EDDK/14L F330 PODIP8F PODIP Y867 WRB N850 ESTAD L23 HAM P605 GESKA
Z711 MONAK MONAK2A EKCH/04R
Steps you've made to test the Pull Request:
1. Programmed FMC
2. Check Flight Plan first waypoint which is BADGO:
grafik
3. Checked first waypoint after dep.:
grafik

Flightnotes:
The Last waypoint, which is the Runway 04R at EKCH has FL330. Is this also related to this PR?
grafik

Issues: Found the one in the flight notes, but maybe not related to the PR 
Overall Rating: If the Issue is not related to the PR - Good, otherwise to be improved

**Comment by Saschl** [comment]
I think I didn’t change anything in that direction but I’ll verify the behaviour with the current dev version. Do you know by any chance what the value should actually be for the rw waypoint? Runway elevation maybe?

**Comment by MisterChocker** [comment]
> Runway elevation maybe

It might be runway threshold elevation

**Comment by Saschl** [comment]
The waypoint shouldn't have the cruise altitude anymore. It now behaves like the current dev version. I tried to improve by setting the runway elevation but in the end I guess it would interfere with the complete rewrite of the predictions that is also in progress.

**Comment by BritishAvgeek** [comment]
> Category: QA Tester
> Name: Noxlee #7727
> Date of testing: Current date 09/12/2020
> Version of the sim: 1.11.7.0
> PR Tested: #2267
> Object of test: Flight plan
> 
> Testing technique: Test Flight -EDDK/14L F330 PODIP8F PODIP Y867 WRB N850 ESTAD L23 HAM P605 GESKA
> Z711 MONAK MONAK2A EKCH/04R
> Steps you've made to test the Pull Request:
> 
> 1. Programmed FMC
> 2. Check Flight Plan first waypoint which is BADGO:
>    grafik
> 3. Checked first waypoint after dep.:
>    grafik
> 
> Flightnotes:
> The Last waypoint, which is the Runway 04R at EKCH has FL330. Is this also related to this PR?
> grafik
> 
> Issues: Found the one in the flight notes, but maybe not related to the PR
> Overall Rating: If the Issue is not related to the PR - Good, otherwise to be improved

That's been an issue for months now nothing to do with this guy's PR... The airports regularly have the cruise alt displayed as a bug in the sim for some reason

**Review by wpine215 — APPROVED** [review]

**Pull request merged**